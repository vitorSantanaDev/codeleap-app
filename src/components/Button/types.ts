import { ButtonHTMLAttributes, ReactNode } from 'react'

export type ButtonBgColor = 'primary' | 'green' | 'red' | 'white'

export type ButtonProps = {
  children: ReactNode
  minimal?: boolean
  bgColor?: ButtonBgColor
} & ButtonHTMLAttributes<HTMLButtonElement>
