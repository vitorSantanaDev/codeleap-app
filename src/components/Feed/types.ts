import { Post } from 'components/PostCard/types'

export type FeedProps = {
  posts: Post[]
  isRefetchingPostsData?: boolean
}
