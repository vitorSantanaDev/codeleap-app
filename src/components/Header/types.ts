import { HTMLAttributes, ReactNode } from 'react'

export type HeaderProps = {
  children: ReactNode
} & HTMLAttributes<HTMLElement>
