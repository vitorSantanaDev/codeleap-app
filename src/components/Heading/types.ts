import { ReactNode } from 'react'

export type HeadingProps = {
  children: ReactNode
  color?: 'white' | 'black' | 'gray'
  size?: 'medium' | 'xlarge'
}
