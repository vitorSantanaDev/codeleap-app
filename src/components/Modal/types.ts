import { ReactNode } from 'react'

export type ModalProps = {
  children: ReactNode
  isVisible: boolean
}
