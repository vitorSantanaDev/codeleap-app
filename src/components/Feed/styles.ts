import styled from 'styled-components'

export const Wrapper = styled.div``

export const LoadingMorePostsWrapper = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`
