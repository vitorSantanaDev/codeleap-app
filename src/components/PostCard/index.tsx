import React from 'react'
import { formatISO } from 'date-fns'
import { useDispatch } from 'react-redux'

import { PostCardProps } from './types'

import Header from 'components/Header'

import EditIcon from 'assets/icons/bx_bx-edit.svg'
import TrashIcon from 'assets/icons/delete-icon.svg'

import { toggleEditPostModal } from 'redux/actions/edit-post-slice'
import { rangeOfTimePassedBetweenTwoDates } from 'utils/formatDate'
import { toggleDeletePostAlert } from 'redux/actions/delete-post-slice'

import * as S from './styles'

const PostCard: React.FC<PostCardProps> = ({ myPost, post }) => {
  const dispatch = useDispatch()

  const postTimeLabel = rangeOfTimePassedBetweenTwoDates(
    post.created_datetime,
    formatISO(new Date())
  )

  return (
    <S.Wrapper>
      <Header>
        <S.HeaderContent>
          <p>{post.title}</p>
          {!!myPost && (
            <S.ActionButtonsWrapper>
              <img
                role="button"
                src={TrashIcon}
                alt="Trash icon"
                aria-label="Trash icon"
                onClick={() => dispatch(toggleDeletePostAlert(post.id))}
              />
              <img
                role="button"
                src={EditIcon}
                alt="Edit icon"
                aria-label="Edit icon"
                onClick={() => dispatch(toggleEditPostModal(post.id))}
              />
            </S.ActionButtonsWrapper>
          )}
        </S.HeaderContent>
      </Header>
      <S.Content>
        <S.PostInfoWrapper>
          <S.UserNameText>@{post.username}</S.UserNameText>
          <S.PostTimeText>{postTimeLabel}</S.PostTimeText>
        </S.PostInfoWrapper>
        <S.Description>{post.content}</S.Description>
      </S.Content>
    </S.Wrapper>
  )
}

export default PostCard
