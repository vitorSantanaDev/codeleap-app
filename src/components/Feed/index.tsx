import React, { forwardRef, useMemo } from 'react'

import theme from 'styles/theme'

import { FeedProps } from './types'
import PostCard from 'components/PostCard'
import { useStateSelector } from 'hooks/useStateSelector'

import InfinityScrollLoading from 'components/InfinityScrollLoading'

import * as S from './styles'

const Feed: React.ForwardRefRenderFunction<HTMLDivElement, FeedProps> = (
  { posts, isRefetchingPostsData },
  ref
) => {
  const user = useStateSelector((state) => state.user.username)

  const postsSortedByMostRecent = useMemo(
    () =>
      posts.sort((a, b) => {
        const createdAtItemA = new Date(a.created_datetime).getTime()
        const createdAtItemB = new Date(b.created_datetime).getTime()
        return createdAtItemB - createdAtItemA
      }),
    [posts]
  )

  return (
    <S.Wrapper>
      {postsSortedByMostRecent.map((post, index) => (
        <div
          key={`${post.id}-${index}`}
          style={{ marginBottom: theme.spacings.small }}
        >
          <PostCard myPost={user === post.username} post={post} />
        </div>
      ))}
      {!isRefetchingPostsData ? (
        <S.LoadingMorePostsWrapper ref={ref} id="loadingMorePosts">
          <InfinityScrollLoading />
        </S.LoadingMorePostsWrapper>
      ) : null}
    </S.Wrapper>
  )
}

export default forwardRef(Feed)
