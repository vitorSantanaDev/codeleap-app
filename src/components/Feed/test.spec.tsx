import React from 'react'
import { screen } from '@testing-library/react'

import { renderWithProviders } from 'utils/render-with-providers'
import { Post } from 'components/PostCard/types'

import Feed from '.'

const postsMock: Post[] = [
  {
    content:
      'Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.',
    title: 'My First Post at CodeLeap Network!',
    created_datetime: new Date().toISOString(),
    id: 1,
    username: 'Cassia'
  },
  {
    content:
      'Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.',
    title: 'My First Post at CodeLeap Network!',
    created_datetime: new Date().toISOString(),
    id: 1,
    username: 'Fran'
  },
  {
    content:
      'Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.',
    title: 'My First Post at CodeLeap Network!',
    created_datetime: new Date().toISOString(),
    id: 1,
    username: 'vitor'
  }
]

describe('<Feed />', () => {
  it('should render posts', () => {
    renderWithProviders(<Feed posts={postsMock} />)

    const posts = screen.getAllByText(/My First Post at CodeLeap Network!/i)
    expect(posts).toHaveLength(3)
  })

  it('should macth snapshots', () => {
    const { container } = renderWithProviders(<Feed posts={postsMock} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
