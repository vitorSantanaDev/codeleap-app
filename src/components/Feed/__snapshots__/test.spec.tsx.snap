// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feed /> should macth snapshots 1`] = `
.c2 {
  font-weight: 700;
  color: #FFFFFF;
  padding: calc(2.4rem + 0.3rem) calc(3.2rem + 0.5rem);
  background-color: #7695EC;
  font-size: calc(2.0rem + 0.2rem);
}

.c3 {
  width: 100%;
  margin: 0 auto;
  max-width: 130rem;
}

.c0 {
  border: 1px solid #999999;
  border-radius: 0.8rem;
  border-radius: 0.8rem;
}

.c0 .c1 {
  border-top-right-radius: 0.8rem;
  border-top-left-radius: 0.8rem;
}

.c4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  width: 100%;
  padding: 2.4rem;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 1.6rem;
}

.c7 {
  font-weight: 700;
  color: #777777;
  font-size: calc(1.6rem + 0.2rem);
}

.c8 {
  font-weight: 400;
  color: #777777;
  font-size: calc(1.6rem + 0.2rem);
}

.c9 {
  color: #000000;
  font-weight: 400;
  font-size: calc(1.6rem + 0.2rem);
}

.c11 {
  display: inline-block;
  position: relative;
  width: 8rem;
  height: 8rem;
}

.c11 div {
  position: absolute;
  top: 3.3rem;
  width: 1.3rem;
  height: 1.3rem;
  border-radius: 50%;
  background: #7695EC;
  -webkit-animation-timing-function: cubic-bezier(0,1,1,0);
  animation-timing-function: cubic-bezier(0,1,1,0);
}

.c11 div:nth-child(1) {
  left: 0.8rem;
  -webkit-animation: lds-ellipsis1 0.6s infinite;
  animation: lds-ellipsis1 0.6s infinite;
}

.c11 div:nth-child(2) {
  left: 0.8rem;
  -webkit-animation: lds-ellipsis2 0.6s infinite;
  animation: lds-ellipsis2 0.6s infinite;
}

.c11 div:nth-child(3) {
  left: 3.2rem;
  -webkit-animation: lds-ellipsis2 0.6s infinite;
  animation: lds-ellipsis2 0.6s infinite;
}

.c11 div:nth-child(4) {
  left: 5.6rem;
  -webkit-animation: lds-ellipsis3 0.6s infinite;
  animation: lds-ellipsis3 0.6s infinite;
}

.c10 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

<div
  class=""
>
  <div
    style="margin-bottom: 2.4rem;"
  >
    <div
      class="c0"
    >
      <header
        class="c1 c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <p>
              My First Post at CodeLeap Network!
            </p>
          </div>
        </div>
      </header>
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            @
            Cassia
          </h2>
          <span
            class="c8"
          >
            0 seconds ago
          </span>
        </div>
        <p
          class="c9"
        >
          Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.
        </p>
      </div>
    </div>
  </div>
  <div
    style="margin-bottom: 2.4rem;"
  >
    <div
      class="c0"
    >
      <header
        class="c1 c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <p>
              My First Post at CodeLeap Network!
            </p>
          </div>
        </div>
      </header>
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            @
            Fran
          </h2>
          <span
            class="c8"
          >
            0 seconds ago
          </span>
        </div>
        <p
          class="c9"
        >
          Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.
        </p>
      </div>
    </div>
  </div>
  <div
    style="margin-bottom: 2.4rem;"
  >
    <div
      class="c0"
    >
      <header
        class="c1 c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <p>
              My First Post at CodeLeap Network!
            </p>
          </div>
        </div>
      </header>
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <h2
            class="c7"
          >
            @
            vitor
          </h2>
          <span
            class="c8"
          >
            0 seconds ago
          </span>
        </div>
        <p
          class="c9"
        >
          Curabitur suscipit suscipit tellus. Phasellus consectetuer vestibulum elit. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.
        </p>
      </div>
    </div>
  </div>
  <div
    class="c10"
    id="loadingMorePosts"
  >
    <div
      class="c11"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;
